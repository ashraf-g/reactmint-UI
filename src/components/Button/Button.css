@import '../../styles/index.css';

.button {
  box-sizing: border-box;
  display: inline-block;
  padding: var(--button-padding, 0.5rem 1rem);
  border-radius: var(--button-border-radius, 0.375rem);
  font-family: var(--mint-font-family, inherit);
  font-size: 1rem;
  cursor: pointer;
  transition: background var(--transition-fast), color var(--transition-fast);
  border: 1px solid transparent;
  background: transparent;
  color: var(--text-primary);
}

.button--primary {
  background-color: var(--color-primary);
  color: white;
  border-color: var(--color-primary);
}

.button--primary:hover {
  background-color: var(--color-primary-hover);
}

.button--secondary {
  background-color: var(--color-secondary);
  color: white;
  border-color: var(--color-secondary);
}

.button--secondary:hover {
  background-color: var(--color-secondary-hover);
}

.button--tertiary {
  background-color: var(--color-tertiary);
  color: white;
  border-color: var(--color-tertiary);
}

.button--tertiary:hover {
  background-color: var(--color-tertiary-hover);
}

.button--outline {
  background: transparent;
  color: var(--text-primary);
  border-color: var(--border-primary);
}

.button--outline:hover {
  background-color: var(--surface-primary);
}

.button--ghost {
  background: transparent;
  color: var(--text-secondary);
  border-color: var(--border-secondary);
}

.button--ghost:hover {
  background-color: var(--surface-secondary);
}

.button--success {
  background-color: var(--color-success);
  color: white;
  border-color: var(--color-success);
}

.button--success:hover {
  background-color: var(--color-success-hover);
}

.button--warning {
  background-color: var(--color-warning);
  color: white;
  border-color: var(--color-warning);
}

.button--warning:hover {
  background-color: var(--color-warning-hover);
}

.button--danger {
  background-color: var(--color-error);
  color: white;
  border-color: var(--color-error);
}

.button--danger:hover {
  background-color: var(--color-error-hover);
}

.button--link {
  background: transparent;
  color: var(--text-link);
  text-decoration: underline;
  border: none;
}

.button--link:hover {
  color: var(--text-link-hover);
}

/* Default style */
.button--default {
  border: 1px solid var(--border-primary);
  background: var(--surface-primary);
  color: var(--text-primary);
}

/* Disabled state */
.button[disabled],
.button[aria-disabled="true"] {
  opacity: 0.6;
  cursor: not-allowed;
  pointer-events: none;
}