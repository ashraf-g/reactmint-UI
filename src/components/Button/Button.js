import React from "react";
import "./Button.css";

const Button = ({
  children,
  className = "",
  variant = "default",
  disabled = false,
  ...props
}) => {
  const variantClass = `button--${variant}`;

  return (
    <button
      className={`button ${variantClass} ${className}`.trim()}
      data-testid="button"
      disabled={disabled}
      {...props}
    >
      {children}
    </button>
  );
};

export default Button;
