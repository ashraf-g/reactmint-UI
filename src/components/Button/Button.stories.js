import React from "react";
import Button from "./Button";
import "./Button.css";

export default {
  title: "Components/Button",
  component: Button,
  tags: ["autodocs"],
  argTypes: {
    variant: {
      control: {
        type: "select",
        labels: {
          default: "Default",
          primary: "Primary",
          secondary: "Secondary",
        },
      },
      options: ["default", "primary", "secondary"],
    },
    children: {
      control: "text",
    },
    onClick: {
      action: "clicked",
    },
  },
  args: {
    children: "Click me",
  },
  parameters: {
    layout: "padded",
    backgrounds: {
      default: "light",
      values: [
        { name: "light", value: "#ffffff" },
        { name: "dark", value: "#333333" },
      ],
    },
  },
};

const Template = (args) => <Button {...args} />;

export const Default = Template.bind({});

export const Primary = Template.bind({});
Primary.args = {
  variant: "primary",
  children: "Primary Action",
};

export const Secondary = Template.bind({});
Secondary.args = {
  variant: "secondary",
  children: "Secondary Action",
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
  children: "Disabled Button",
};
