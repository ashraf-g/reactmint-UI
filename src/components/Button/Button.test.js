import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Button from "./Button";

describe("Button Component", () => {
  // 1. Basic Rendering
  it("renders without crashing", () => {
    render(<Button />);
    expect(screen.getByRole("button")).toBeInTheDocument();
  });

  it("renders children text correctly", () => {
    render(<Button>Click Me</Button>);
    expect(screen.getByRole("button")).toHaveTextContent("Click Me");
  });

  it("renders children elements correctly", () => {
    render(
      <Button>
        <span>Icon</span> Text
      </Button>
    );
    expect(screen.getByRole("button")).toContainHTML("<span>Icon</span> Text");
  });

  // 2. Class Name Handling
  it("applies default class", () => {
    render(<Button />);
    expect(screen.getByRole("button")).toHaveClass("button");
  });

  it("merges custom className correctly", () => {
    render(<Button className="custom-class" />);
    const button = screen.getByRole("button");
    expect(button).toHaveClass("button");
    expect(button).toHaveClass("custom-class");
  });

  // 3. Variant Handling
  it("applies default variant when none specified", () => {
    render(<Button />);
    const button = screen.getByRole("button");
    expect(button).toHaveClass("button");
    expect(button).not.toHaveClass("button--primary");
    expect(button).not.toHaveClass("button--secondary");
  });

  it("applies primary variant correctly", () => {
    render(<Button variant="primary" />);
    expect(screen.getByRole("button")).toHaveClass("button--primary");
  });

  it("applies secondary variant correctly", () => {
    render(<Button variant="secondary" />);
    expect(screen.getByRole("button")).toHaveClass("button--secondary");
  });

  it("handles unknown variant gracefully", () => {
    render(<Button variant="unknown" />);
    const button = screen.getByRole("button");
    expect(button).toHaveClass("button");
    expect(button).not.toHaveClass("button--unknown");
  });

  // 4. Props Handling
  it("passes all other props to the button element", () => {
    const onClick = jest.fn();
    render(
      <Button
        id="test-button"
        data-test="test-data"
        aria-label="Test Button"
        onClick={onClick}
      />
    );

    const button = screen.getByRole("button");
    fireEvent.click(button);

    expect(button).toHaveAttribute("id", "test-button");
    expect(button).toHaveAttribute("data-test", "test-data");
    expect(button).toHaveAttribute("aria-label", "Test Button");
    expect(onClick).toHaveBeenCalled();
  });

  it("handles click events", () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick} />);
    fireEvent.click(screen.getByRole("button"));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  // 5. Accessibility - Removed redundant role test since native button has implicit role
  it("accepts aria-* props", () => {
    render(<Button aria-disabled="true" aria-label="Submit" />);
    const button = screen.getByRole("button");
    expect(button).toHaveAttribute("aria-disabled", "true");
    expect(button).toHaveAttribute("aria-label", "Submit");
  });

  // 6. Edge Cases
  it("renders correctly with no children", () => {
    render(<Button />);
    expect(screen.getByRole("button")).toBeEmptyDOMElement();
  });

  it("combines multiple classNames correctly", () => {
    render(<Button className="class1 class2" variant="primary" />);
    const button = screen.getByRole("button");
    expect(button).toHaveClass("button");
    expect(button).toHaveClass("button--primary");
    expect(button).toHaveClass("class1");
    expect(button).toHaveClass("class2");
  });

  // 7. Updated Snapshot Tests
  it("matches default button snapshot", () => {
    const { asFragment } = render(<Button />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches primary variant snapshot", () => {
    const { asFragment } = render(<Button variant="primary">Primary</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches button with custom class snapshot", () => {
    const { asFragment } = render(<Button className="custom">Custom</Button>);
    expect(asFragment()).toMatchSnapshot();
  });
});