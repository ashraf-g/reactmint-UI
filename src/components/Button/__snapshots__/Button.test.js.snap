// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component matches button with custom class snapshot 1`] = `
<DocumentFragment>
  <button
    class="button  custom"
    data-testid="button"
    role="button"
  >
    Custom
  </button>
</DocumentFragment>
`;

exports[`Button Component matches default button snapshot 1`] = `
<DocumentFragment>
  <button
    class="button"
    data-testid="button"
    role="button"
  />
</DocumentFragment>
`;

exports[`Button Component matches primary variant snapshot 1`] = `
<DocumentFragment>
  <button
    class="button button--primary"
    data-testid="button"
    role="button"
  >
    Primary
  </button>
</DocumentFragment>
`;
